File: 2015/05.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
require 'set'

def nice(str)
  vowels = Set.new(['a','e','i','o','u',])
  bad    = ['ab','cd','pq','xy']

  return false if bad.any? { |b| str.include? b }

  count = str.chars.count { |c| vowels.member? c }
  return count > 2 && str =~ /(.)\1/ # matches a repeating character
end

# gotta love regex
# first matches a pair of letters that appears twice
# second matches a letters that repeats with one letter in between
def nice2(str)
  return str =~ /(..).*\1/ && str =~ /(.).\1/
end

input  = File.readlines('input/05.txt').map &:chomp
silver = input.count { |x| nice  x }
gold   = input.count { |x| nice2 x }

puts(
  "Day 05\n"       \
  "======\n"       \
  "✮: #{silver}\n" \
  "★: #{gold}"
)

File: 2015/05.clj
(require '[clojure.string :refer (includes? split-lines)])

(def lines (split-lines (slurp "input/05.txt")))

(defn nice [str]
  (let [vowels #{\a \e \i \o \u}
        bad ["ab" "cd" "pq" "xy"]]
    (and (not-any?  #(includes? str %) bad)
         (< 2 (count (filter #(contains? vowels %) str)))
         (boolean (re-find #"(.)\1" str)))))

(defn nice2 [str]
  (every? #(boolean (re-find % str))
          [#"(..).*\1"  #"(.).\1"]))

(run! println (map #(count (filter % lines)) [nice nice2]))